packages/test/test-gc-sweep-tests/src/test/gcSweepTest.spec.ts

```diff
@@ -0,0 +1,355 @@
+        if (connectedContainers.length <= 0) {
+            return;
+        }
+        await createDataStoreForContainer(getRandomContainer());
+    };
+
+    const getTestNode = (id: string): TestNode => {
+        const testNode = testNodes.get(id);
+        assert(testNode !== undefined, `Datastore ${id} should exist in datastores!`);
+        return testNode;
+    };
+
+    const updateReferencedDataStores = (id: string) => {
+        if (!refDataStores.has(id)) {
+            refDataStores.add(id);
+            unrefDataStores.delete(id);
+            for (const childId of getTestNode(id).children) {
+                updateReferencedDataStores(childId);
+            }
+        }
+    };
+
+    const hasReferencedParent = (id: string): boolean => {
+        for (const parentNode of getTestNode(id).parents) {
+            if (refDataStores.has(parentNode)) {
+                return true;
+            }
+        }
+        return false;
+    };
+
+    const updateUnreferencedDataStores = (id: string) => {
+        if (!unrefDataStores.has(id) && !hasReferencedParent(id)) {
+            unrefDataStores.add(id);
+            refDataStores.delete(id);
+            const testNode = getTestNode(id);
+            for (const childId of testNode.children) {
+                updateUnreferencedDataStores(childId);
+            }
+        }
+    };
+
+    const referenceDataStore = (dataStore: TestDataObject, parentDataStore: TestDataObject) => {
+        parentDataStore._root.set(dataStore.id, dataStore.handle);
+        getTestNode(parentDataStore.id).children.add(dataStore.id);
+        getTestNode(dataStore.id).parents.add(parentDataStore.id);
+        if (refDataStores.has(parentDataStore.id)) {
+            updateReferencedDataStores(dataStore.id);
+        }
+    };
+
+    const unreferenceDataStore = (dataStore: TestDataObject, parentDataStore: TestDataObject) => {
+        if (rootDataStores.has(dataStore.id)) {
+            return;
+        }
+        parentDataStore._root.delete(dataStore.id);
+        getTestNode(parentDataStore.id).children.delete(dataStore.id);
+        getTestNode(dataStore.id).parents.delete(parentDataStore.id);
+        updateUnreferencedDataStores(dataStore.id);
+    };
+
+    const getRandomContainer = (): IContainer => {
+        const id = randomInteger(0, connectedContainers.length - 1);
+        return connectedContainers[id];
+    };
+
+    const getRandomDataStoreFromContainer = async (container: IContainer) => {
+        const random = randomInteger(0, testNodes.size - 1);
+        const dataStoreId = Array.from(testNodes.values())[random].id;
+        return requestFluidObject<TestDataObject>(container, dataStoreId);
+    };
+
+    const getRandomParentForDataStore = async (dataStore: TestDataObject, container: IContainer) => {
+        const parents = getTestNode(dataStore.id).parents;
+        const random = randomInteger(0, parents.size - 1);
+        const parentId = Array.from(parents.values())[random];
```

Comment:
Doesn't look like this function handles gracefully the case where there are no parents.